### PR TITLE
packaging: disable -buildmode=pie on all arches

### PR DESCRIPTION
**Pull request by mvo5**
Building go binaries with -buildmode=pie causes issue on ARM
32bit boards. This was originally reported by a customer and
tracked in LP #1822738.

The issue is that the new GC in 1.7+ allocates a large chunk
of memory (>150Mb in our case) on ARM systems when using
buildmode=pie on the low-memory system where this was observed.

The following happens:
1. We build snap/snapd with -buildmode=pie
2. There is a bug in the malloc.go code in 1.10
3. Because the board this runs on has too little ram the kernel
   relocates it from a relatively low memory start address to a pretty
   high one. Because the GO memory system assumes it initially owns
   all the memory from the point where it start to the start it creates
   a huge GC bitmap memory region.
4. Because the pi2 does have enough ram the kernel does not relocate it
   so we cannot observe the error there.

Using -buildmode=pie is not super important on GO according to
one of the authors:
https://groups.google.com/forum/#!topic/golang-nuts/Jd9tlNc6jUE

The relevant LP bug:
https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/1822738

### CODE REVIEW

**Comment by zyga** [comment]
I'm marking it as blocked because build mode pie is required for build-id, which snapd depends on.

**Comment by mvo5** [comment]
@zyga The `tests/main/system-key` test will check if we have a system-key:
```
MATCH '"build-id":"[0-9a-z]+"' < /var/lib/snapd/system-key
```
for snapd. So lets see if that is successful or fails. On Ubuntu it seems we do have a valid build-id even without -buildmode=pie with go-1.10.

**Comment by bboozzoo** [comment]
Since we were able to reproduce this with a piece of most trivial Go code, we did a rundown of Go versions. We've verified 1.7 and 1.10 to fail, 1.11 was ok. I've bisected Go versions between 1.10beta2 and 1.11beta1, found this commit https://github.com/golang/go/commit/c0392d2e7fbdcd38aafb959e94daf6bbafe2e4e9 to be the first one when the problem is 'fixed' (or not triggered on that particular kernel + userland combination). The observed `mmap()` calls request a sane amount of memory, 262kB instead of 150MB like the broken versions did.

**Comment by jdstrand** [comment]
I've not verified the veracity of the assertion that PIE is not useful for go programs (perhaps @mwhudson has thoughts; if not, no worries), but perhaps someone can get the 1.11 fixes backported into Ubuntu? This sounds like a potential problem for all golang users....

**Comment by mwhudson** [comment]
I think I would agree with Russ that PIE for Go is mostly a defence against being linked with buggy C code and as snapd only links to libc we're probably not on super shaky ground there.

I'm not sure fwiw why building with Go 1.6 does not see this problem. It has the same behaviour of allocating a bitmap from 0 to the the end of the heap bitmap as far as I can see.

As to backporting the fixes, that commit touches a very core area of the Go runtime. I've found that if I include two other commits first (c690dc8155, bb29700bbe) then the referenced commit can be cherry-picked but really the safer option would be backporting Go 1.11 (or 1.12) and using that instead. 

In summary +1 from me on disabling PIE for arm until you are building with Go 1.11 or better. But I've just seen that this PR disables PIE on all architectures -- that seems excessive. This is a 32-bit-only problem (and PIE is already disabled on i386, thanks to other bugs that are probably my fault).

**Comment by alexmurray** [comment]
I agree with @mwhudson - this should disable PIE on arm (32-bit - so armhf) ONLY - and then there should be a clear expectation this is reverted once 1.11 is available (unless there is evidence this also occurs on other archs).

**Comment by mvo5** [comment]
Thanks for the feedback.  As mentioned we already had to disable -buildmode=pie on i386 because of LP: #1711052 -I' m worried that -buildmode=pie is not something that is considered important by the GO upstream team and hence its (much?) less well tested than the regular build mode.

So given that:
a) buildmode=pie does not give us much (any?) extra protection (according to upstream)
b) its is risky (because upstream considers it less important, hence less testing)
c) it broke devices in the field and the recovery for those will be painful (broken IoT devices in houses/at random place without a working snapd are really hard to fix)

I'm really worried about the next incident when we need to disable PIE for the next architecture. To me the risk of breakage amd64/aarch64 in some yet unknown way and to have another class of devices broken seems to outweigh the benefits of PIE. Maybe its the fresh memory of debugging this (and explaining it to affected people) but at this point I feel not confident in -buildmode=pie anymore.

**Comment by pedronis** [comment]
There is overall disagreement on the exact path forward, and no regression test in place. Marking blocked for now.

**Comment by jdstrand** [comment]
Please note that another member of the security team (Seth Arnold) expressed an opinion on the bug:

"
> Proposed tenative fix in snapd is to disable PIE builds. Relevant PR:
> https://github.com/snapcore/snapd/pull/6700
I dislike this change.

While ASLR is not particularly strong on 32 bit platforms, it is
significantly more useful when used with 64 bit platforms. Go itself may
be relatively safe from the common mistakes that make ASLR really useful
in C, but loading C libraries does happen.

Furthermore I'm not confident it will actually help. The problem appears
to be that a specific size was requested at a specific location and
MAP_FIXED prevents the kernel from finding another region where the
requested allocation could fit:

mmap2(0x7f900000, 150536192, PROT_READ|PROT_WRITE,
MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = -1 ENOMEM (Cannot
allocate memory)

Removing the MAP_FIXED would probably allow this allocation to continue,
albeit at a different location, and would also avoid the caveat given
in the mmap(2) manpage:


              Furthermore, this option is extremely hazardous (when
              used on its own), because it forcibly removes
              preexisting mappings, making it easy for a
              multithreaded process to corrupt its own address
              space.
"

**Comment by mwhudson** [comment]
> I'm really worried about the next incident when we need to disable PIE for the next architecture. To me the risk of breakage amd64/aarch64 in some yet unknown way and to have another class of devices broken seems to outweigh the benefits of PIE. Maybe its the fresh memory of debugging this (and explaining it to affected people) but at this point I feel not confident in -buildmode=pie anymore.

I can't tell you that you are wrong here, but I will say that the issue here is very different from the i386 one. That's a codegen bug, this one is a misfeature with the GC that bumps into overcommit limits on 32 bit systems when the executable is not loaded at a low base address. Disabling PIE on 32 bit is an entirely reasonable thing to do in response to this issue, but disabling it on 64 bit is less justified because the GC works very differently on 64 bit systems.

I can't promise that there aren't PIE-specific codegen bugs on other architectures but I can promise that the PIE codegen code is much much less bad there (i386 is notoriously bad for PIC codegen).

**Comment by bboozzoo** [comment]
@mvo5 has opened #6706 to disable PIE only for armhf.

**Comment by mvo5** [comment]
We need to resolve this so I opened 6706 - but I am still not convinced that this one here should be closed. Maybe next week when the crisis is over we can have a chat in a hangout with the involved parties.

**Comment by bboozzoo** [comment]
Digging into PIE a bit further. Looks like snap-confine is not a PIE either. This is x86_64:
```
$ scanelf -a /snap/core/current/usr/lib/snapd/snap-confine
 TYPE    PAX   PERM ENDIAN STK/REL/PTL TEXTREL RPATH BIND FILE 
ET_EXEC PeMRxS 6755 LE RW- R-- RW-    -      -   LAZY /snap/core/current/usr/lib/snapd/snap-confine
```
And i386:
```
google:ubuntu-16.04-32 .../tests/main/buildmode# scanelf -a /snap/core/current/usr/lib/snapd/snap-confine
 TYPE    PAX   PERM ENDIAN STK/REL/PTL TEXTREL RPATH BIND FILE 
ET_EXEC PeMRxS 6755 LE RW- R-- RW-    -      -   LAZY /snap/core/current/usr/lib/snapd/snap-confine 
```
Is this something we should adjust?

**Comment by jdstrand** [comment]
> Is this something we should adjust?

Yikes, yes; this setuid C application definitely needs PIE.

**Comment by bboozzoo** [comment]
> Yikes, yes; this setuid C application definitely needs PIE.

I'll look into fixing this.

**Comment by bboozzoo** [comment]
#6706 landed, can this one be closed now?

**Comment by Conan-Kudo** [comment]
@mvo5 @bboozzoo Note that this issue does not apply to Fedora or EL7 builds, as we're using golang 1.11 or newer to build there.

**Comment by mvo5** [comment]
Closing for now but my opinion for this has not changed.
